src/meta/processors/admin/Balancer.cpp

```diff
@@ -0,0 +1,316 @@
+}
+
+std::vector<BalanceTask> Balancer::genTasks(GraphSpaceID spaceId) {
+    CHECK(!!plan_) << "plan should not be nullptr";
+    std::unordered_map<HostAddr, std::vector<PartitionID>> hostParts;
+    int32_t totalParts = 0;
+    getHostParts(spaceId, hostParts, totalParts);
+    if (totalParts == 0 || hostParts.empty()) {
+        LOG(ERROR) << "Invalid space " << spaceId;
+        return std::vector<BalanceTask>();
+    }
+    auto activeHosts = ActiveHostsMan::instance()->getActiveHosts();
+    std::vector<HostAddr> newlyAdded;
+    std::vector<HostAddr> lost;
+    calDiff(hostParts, activeHosts, newlyAdded, lost);
+    decltype(hostParts) newHostParts(hostParts);
+    for (auto& h : newlyAdded) {
+        newHostParts.emplace(h, std::vector<PartitionID>());
+    }
+    for (auto& h : lost) {
+        newHostParts.erase(h);
+    }
+    LOG(INFO) << "Now, try to balance the newHostParts";
+    // We have two parts need to balance, the first one is parts on lost hosts
+    // The seconds one is parts on unbalanced host in newHostParts.
+    std::vector<BalanceTask> tasks;
+    for (auto& h : lost) {
+        auto& lostParts = hostParts[h];
+        for (auto& partId : lostParts) {
+            auto ret = hostWithMinimalParts(newHostParts, partId);
+            if (!ret.ok()) {
+                LOG(ERROR) << "Error:" << ret.status();
+                return std::vector<BalanceTask>();
+            }
+            auto& luckyHost = ret.value();
+            newHostParts[luckyHost].emplace_back(partId);
+            tasks.emplace_back(plan_->id_,
+                               spaceId,
+                               partId,
+                               h,
+                               luckyHost,
+                               kv_,
+                               client_.get());
+        }
+    }
+    if (newHostParts.size() < 2) {
+        LOG(INFO) << "Too few hosts, no need for balance!";
+        return tasks;
+    }
+    balanceParts(plan_->id_, spaceId, newHostParts, totalParts, tasks);
+    return tasks;
+}
+
+void Balancer::balanceParts(BalanceID balanceId,
+                            GraphSpaceID spaceId,
+                            std::unordered_map<HostAddr, std::vector<PartitionID>>& newHostParts,
+                            int32_t totalParts,
+                            std::vector<BalanceTask>& tasks) {
+    float avgLoad = static_cast<float>(totalParts)/newHostParts.size();
+    LOG(INFO) << "The expect avg load is " << avgLoad;
+    int32_t minLoad = std::floor(avgLoad);
+    int32_t maxLoad = std::ceil(avgLoad);
+    auto hosts = sortedHostsByParts(newHostParts);
+    CHECK_GT(hosts.size(), 1);
+    auto maxPartsHost = hosts.back();
+    auto minPartsHost = hosts.front();
+    auto lastDelta = maxPartsHost.second - minPartsHost.second + 1;
+    while (maxPartsHost.second > maxLoad
+            || minPartsHost.second < minLoad
+            || maxPartsHost.second - minPartsHost.second < lastDelta) {
```

Comment:
why is less than lastDelta? If current delta is 0, this function will process the loop, but will do nothing because of “partsFrom.size() <= partsTo.size() + 1”

Reply:
Because i want the delta be more small after each iteration.  And if noActions happened, the loop will be broken.